File: solutions/lua/binary-search/1/binary-search.lua
return function(array, target)
    local low = 1
    local high = #array

    while low <= high do
        local mid = math.floor((low + high) / 2)
        if array[mid] == target then
            return mid
        elseif array[mid] < target then
            low = mid + 1
        else
            high = mid - 1
        end
    end

    return -1
end

File: solutions/c/binary-search/1/binary_search.c
#include "binary_search.h"

const int *binary_search(int value, const int *arr, size_t length) {
    int low = 0, high = length - 1;

    while (low <= high) {
        int index = low + (high - low) / 2;
        if (arr[index] == value) {
            return &arr[index];
        } else if (value < arr[index]) {
            high = index - 1;
        } else {
            low = index + 1;
        }
    }

    return NULL;
}